### ISSUE TITLE
Teleport Operator: `<resource>_labels` does not accept wildcard matcher

### ISSUE THREAD

**Issue by marcoandredinis** [gravitational/teleport#15204]
Expected behavior:
We must be able to set wildcard when configuring Allow rules for any `<resource>_labels` of a Role.

```yaml
kind: role
version: v5
metadata:
  name: interns
spec:
  allow:
    node_labels:
      '*': ['*']
```

Current behavior:
This returns an error:
```
# * <nil>: Invalid value: "The edited file failed validation": ValidationError(TeleportRole.spec.allow.app_labels): invalid type for dev.teleport.resources.v5.TeleportRole.spec.allow.app_labels: got "array", expected "map"
```

Adding another value to the list allows for the object to be correctly loaded
```yaml
    node_labels:
      '*': ['*', 'test']
```
But then Teleport complains of 
```
 "error": "selector *:<val> is not supported"
```
https://github.com/gravitational/teleport/blob/5485498775df97997f71c59b398027366423f268/api/types/role.go#L777

Bug details:
- Teleport version
10.1.2
- Recreation steps
`kubectl create -f ` using the role described above or create the following test in Roles

Or use this test
```go

func TestRoleCreationFromYAML(t *testing.T) {
	ctx := context.Background()

	teleportServer, operatorName := defaultTeleportServiceConfig(t)

	require.NoError(t, teleportServer.Start())

	tClient := clientForTeleport(t, teleportServer, operatorName)
	k8sClient := startKubernetesOperator(t, tClient)

	namespace := createNamespaceForTest(t, k8sClient)
	roleName := "interns"

	roleYAML := yaml.NewYAMLOrJSONDecoder(bytes.NewBuffer([]byte(`
kind: role
version: v5
metadata:
  name: interns
spec:
  allow:
    node_labels:
      '*': ['*']
`)), 1024)

	var roleX resourcesv5.TeleportRole
	err := roleYAML.Decode(&roleX)
	require.NoError(t, err)

	roleX.Namespace = namespace.Name
	err = k8sClient.Create(ctx, &roleX)
	require.NoError(t, err)

	fastEventually(t, func() bool {
		tRole, err := tClient.GetRole(ctx, roleName)
		if trace.IsNotFound(err) {
			return false
		}
		require.NoError(t, err)

		require.Equal(t, tRole.GetName(), roleName)

		require.Contains(t, tRole.GetMetadata().Labels, types.OriginLabel)
		require.Equal(t, tRole.GetMetadata().Labels[types.OriginLabel], types.OriginKubernetes)

		return true
	})

	// The role is deleted in K8S
	k8sDeleteRole(ctx, t, k8sClient, roleName, namespace.Name)

	fastEventually(t, func() bool {
		_, err := tClient.GetRole(ctx, roleName)
		return trace.IsNotFound(err)
	})
}
```
- Debug logs
See above

**Comment by hugoShaka** [comment]
This issue applies to all fields of type `utils.Strings`, regardless of their value.

During CRD generation we define those fields as array of strings in the OpenAPI schema.

However when unmarshalled, `utils.Strings` can either return an array of strings or a single string. When the latter happens, we are going to hit a wall next time the operator wants to update the resource, trying to push a string instead of an array of strings.

CRDs don't support multiple types since v1 (except the int-or-string, but it is behind a specific flag).

Bad (and very bad) ideas to fix this:

- Modify the CRD generation to stop validating when using the type `utils.Strings`
- Modify `utils.Strings` marshalling to be consistent type (no need to touch unmarshalling)
- maintain/build a list of fields to correct after marshalling and add a post-marshalling step
- give the operator its own go modules and `go mod replace` with a copy of utils/strings always returning arrays
- add a virtual array element to ensure we get an array (could be a utils.Strings constant, to avoid the widcard issue mentioned bt marco)

**Comment by hugoShaka** [comment]
I don't think we can change the `utils.Strings` behaviour (correct me if I'm wrong) so the two most promising approaches would be:
- loosening the CRD schema to allow those mixed-type fields
-  post-processing after marshalling

The latter either implies maintaining a list of known fields to process (this can get out of date and bite us again) or walking through the JSONSchema (I'm not sure we have a way to get all the info we need in `TeleportRole.Marshall()`).

I went for the CRD schema loosening and the test is now passing. I'd like other opinions on this approach before going further: @marcoandredinis @r0mant @tigrato

**Comment by marcoandredinis** [comment]
Changing the Marhsal is a bit risky given how many places it's being used.
Although, in theory, the code should always support multiple strings
Not supporting is a bug.

I think it's worth the investigation but if we find it too risky/actually has issues, I think we can follow the proposal of the PR you just created

**Comment by tigrato** [comment]
Why can't we fix the Marshal for `utils.Strings`? 
For me, it's super odd that an array of strings can be represented as a single string if it has only a single element since it's not a valid YAML representation.
As long as the unmarshal is still compatible with a single string, I think it shouldn't be problematic because I think we do not do anything with the YAML representation

What do you think @r0mant ?

**Comment by hugoShaka** [comment]
I tried the change in `utils.Strings` and the integration tests are not more broken than usual. It sounds sane to be permissive with what we unmarshall (input) and strict with what we marshall (output). @r0mant was affraid to touch change the utils behaviour as its usage spreads across all code base.

I'll create a second draft PR and see if I can get all tests to pass

**Comment by r0mant** [comment]
@hugoShaka @tigrato @marcoandredinis Passing tests is not a very good indicator - utils.Strings is used in various places including user-facing configuration and so on so any change to it's marshal/unmarshal behavior will potentially result in backwards incompatible behavior for users. It may be ok or may not be ok, we'd need to research to understand for sure and I don't think this problem warrants the time spent on it for now.

**Comment by marcoandredinis** [comment]
:+1: 

I don't think this will cause any issues because Teleport validates the input even if K8S doesn't

**Comment by hugoShaka** [comment]
It ended up being a bit more complex than that because we use a typed-client, hence input validation de-facto happen when unmarshalling and this crashes the operator. I tried something by switching to unstructured client, this complexifies a bit the process but seems to work. I Should have something in review by EOD.

**Comment by campanamarco** [comment]
+1 for new customer running into this issue @r0mant @marcoandredinis

**Comment by hugoShaka** [comment]
@campanamarco this issue should be already fixed in [10.1.9](https://github.com/gravitational/teleport/releases/tag/v10.1.9). If they are still running into the issue can you provide the version they are running and details about the issue they are facing?
